procrastinate/retry.py

```diff
@@ -6,28 +6,93 @@
 from __future__ import annotations
 
 import datetime
-from typing import Iterable, Union
+import warnings
+from typing import Iterable, Union, overload
 
 import attr
 
-from procrastinate import exceptions, utils
+from procrastinate import exceptions, types, utils
+from procrastinate.jobs import Job
+
+
+class RetryDecision:
+    retry_at: datetime.datetime | None = None
+    priority: int | None = None
+    queue: str | None = None
+    lock: str | None = None
+
+    @overload
+    def __init__(
+        self,
+        retry_at: None = None,
+        retry_in: types.TimeDeltaParams | None = None,
+        priority: int | None = None,
+        queue: str | None = None,
+        lock: str | None = None,
+    ) -> None: ...
+    @overload
+    def __init__(
+        self,
+        retry_at: datetime.datetime | None = None,
+        retry_in: None = None,
+        priority: int | None = None,
+        queue: str | None = None,
+        lock: str | None = None,
+    ) -> None: ...
+    def __init__(
+        self,
+        retry_at: datetime.datetime | None = None,
+        retry_in: types.TimeDeltaParams | None = None,
+        priority: int | None = None,
+        queue: str | None = None,
+        lock: str | None = None,
+    ) -> None:
+        if retry_at and retry_in is not None:
+            raise ValueError("Cannot set both retry_at and retry_in")
+
+        if retry_in is not None:
+            retry_at = utils.datetime_from_timedelta_params(retry_in)
+
+        self.retry_at = retry_at
+        self.priority = priority
+        self.queue = queue
+        self.lock = lock
 
 
 class BaseRetryStrategy:
     """
     If you want to implement your own retry strategy, you can inherit from this class.
-    Child classes only need to implement `get_schedule_in`.
+    Child classes only need to implement `get_retry_decision`.
     """
 
     def get_retry_exception(
-        self, exception: BaseException, attempts: int
+        self, exception: BaseException, job: Job
     ) -> exceptions.JobRetry | None:
-        schedule_in = self.get_schedule_in(exception=exception, attempts=attempts)
-        if schedule_in is None:
-            return None
-
-        schedule_at = utils.utcnow() + datetime.timedelta(seconds=schedule_in)
-        return exceptions.JobRetry(schedule_at.replace(microsecond=0))
+        try:
+            retry_decision = self.get_retry_decision(exception=exception, job=job)
+            if retry_decision is None:
+                return None
+
+            return exceptions.JobRetry(retry_decision=retry_decision)
+        except NotImplementedError as err:
+            try:
+                schedule_in = self.get_schedule_in(
+                    exception=exception, attempts=job.attempts
+                )
+            except NotImplementedError:
+                raise err
```

Comment:
I wonder why coverage tells me that 82-83 is not tested. I would have expected that this is covered by `test_missing_implementation_in_custom_retry_strategy`, or is it because I re-raised the error there?

Reply:
Never mind ... I tested the wrong method. Will fix it.

Reply:
`raise err from None` I believe, otherwise the traceback will be ugly, I think.